### ISSUE TITLE
CFI enabled kernels

### ISSUE THREAD

**Issue by jpds** [QubesOS/qubes-issues#6565]
**Describe the solution you'd like**

Qubes should enable CFI for the Linux kernels used in the dom0 and of course the domUs.

Relevant literature:

- https://lwn.net/Articles/810077/
- https://www.phoronix.com/scan.php?page=news_item&px=Clang-CFI-Linux-Patches
- https://lore.kernel.org/lkml/20210312004919.669614-1-samitolvanen@google.com/
- https://outflux.net/slides/2020/lss/cfi.pdf
- https://android-developers.googleblog.com/2018/10/control-flow-integrity-in-android-kernel.html

**Comment by h01ger** [comment]
On Wed, Apr 28, 2021 at 05:50:30AM -0700, Jonathan Davies wrote:
> Qubes should enable CFI for the Linux kernels used in the dom0 and of course the domUs.

do you happen to know why it's not enabled by default in Debian, Ubuntu, Fedora,
SuSE or Archlinux kernels? Or is it?


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁  holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀  OpenPGP: B8BF54137B09D35CF026FE9D 091AB856069AAA1C
 ⠈⠳⣄

Try to imagine a future where paying for your morning coffee involved smashing
an iPhone and burning enough fossil fuels to run your entire household for 60
days. That's the environmental cost of the "revolutionary" technology behind
Bitcoin in a nutshell. https://twitter.com/smdiehl/status/1350869944888664064

**Comment by jpds** [comment]
@h01ger The patches are just landing in Linux itself, but I filed this as more of a tracking bug.

**Comment by h01ger** [comment]
On Wed, Apr 28, 2021 at 06:36:24AM -0700, Jonathan Davies wrote:
> @h01ger The patches are just landing in Linux itself, but I filed this as more of a tracking bug.

ah, thanks!


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁  holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀  OpenPGP: B8BF54137B09D35CF026FE9D 091AB856069AAA1C
 ⠈⠳⣄

**Comment by marmarek** [comment]
This also requires Xen to expose appropriate features to VMs. This is
in very early stage: ***@***.***/T/#t

**Comment by kees** [comment]
> On Wed, Apr 28, 2021 at 05:50:30AM -0700, Jonathan Davies wrote: Qubes should enable CFI for the Linux kernels used in the dom0 and of course the domUs.
> do you happen to know why it's not enabled by default in Debian, Ubuntu, Fedora, SuSE or Archlinux kernels? Or is it?

Short version: CFI requires Clang, and most distro kernel builds are still very closely tied to GCC. (Also the x86 CFI is still under development, the current series is arm64 only.)

**Comment by h01ger** [comment]
On Thu, Apr 29, 2021 at 09:33:59AM -0700, Kees Cook wrote:
> Short version: CFI requires Clang, and most distro kernel builds are still very closely tied to GCC. (Also the x86 CFI is still under development, the current series is arm64 only.)

oh, ic & thanks for these details, Kees!


-- 
cheers,
	Holger

 ⢀⣴⠾⠻⢶⣦⠀
 ⣾⠁⢠⠒⠀⣿⡁  holger@(debian|reproducible-builds|layer-acht).org
 ⢿⡄⠘⠷⠚⠋⠀  OpenPGP: B8BF54137B09D35CF026FE9D 091AB856069AAA1C
 ⠈⠳⣄

If secure encryption is outlawed, only criminals will have it.

**Comment by DemiMarie** [comment]
> > On Wed, Apr 28, 2021 at 05:50:30AM -0700, Jonathan Davies wrote: Qubes should enable CFI for the Linux kernels used in the dom0 and of course the domUs.
> > do you happen to know why it's not enabled by default in Debian, Ubuntu, Fedora, SuSE or Archlinux kernels? Or is it?
> 
> Short version: CFI requires Clang, and most distro kernel builds are still very closely tied to GCC. Also the x86 CFI is still under development, the current series is arm64 only, but the delta is small if anyone would like to test it:
> https://github.com/samitolvanen/linux/commits/clang-cfi

To add to this: CFI requires LTO, which in turn requires a massive amount of memory (16GiB, IIRC).

**Comment by kees** [comment]
> To add to this: CFI requires LTO, which in turn requires a massive amount of memory (16GiB, IIRC).

Yup, that too, though ThinLTO (the default) should make that WAY better. A reasonably normal kernel config (not, like, allyesconfig) appears to be under 8GiB.

**Comment by marmarek** [comment]
Some related info: the Xen version shipped in R4.1 has CET shadow stack support enabled by default (if hardware supports it). And soon will have BTI support backported too.
